File: src/php/BaseAgent.php
<?php

declare(strict_types=1);

namespace AgentHarness;

use OpenAI;
use OpenAI\Contracts\ClientContract;
use OpenAI\Responses\Chat\CreateResponse;
use OpenAI\Responses\Chat\CreateStreamedResponse;
use OpenAI\Responses\StreamResponse;

class BaseAgent
{
    protected ClientContract $client;

    /**
     * @param array<string, mixed> $completionParams Extra options forwarded to chat()->create()
     * @param ClientContract|null  $client           Inject a pre-built client (e.g. ClientFake) for tests
     */
    public function __construct(
        public readonly string $model,
        public readonly ?string $system = null,
        public readonly int $maxTurns = 20,
        public readonly int $maxRetries = 2,
        public readonly bool $stream = true,
        public readonly ?string $baseUrl = null,
        public readonly ?string $apiKey = null,
        public readonly array $completionParams = [],
        ?ClientContract $client = null,
    ) {
        $this->client = $client ?? $this->buildClient();
    }

    private function buildClient(): ClientContract
    {
        $factory = OpenAI::factory();

        $key = $this->apiKey ?? getenv('OPENAI_API_KEY') ?: 'sk-placeholder';
        $factory = $factory->withApiKey($key);

        if ($this->baseUrl !== null) {
            $factory = $factory->withBaseUri($this->baseUrl);
        }

        return $factory->make();
    }

    /**
     * Extension point for building the system prompt.
     */
    protected function buildSystemPrompt(?string $basePrompt, mixed $context): ?string
    {
        return $basePrompt;
    }

    /**
     * Extension point called at the start of a run.
     */
    protected function onRunStart(mixed $context): void
    {
    }

    /**
     * Extension point called at the end of a run.
     */
    protected function onRunEnd(mixed $context): void
    {
    }

    /**
     * Extension point for handling an LLM response before the agentic loop continues.
     * Return null to signal the loop should stop.
     */
    protected function handleResponse(array $response, RunContext $context): ?array
    {
        return $response;
    }

    /**
     * Consume a streaming chat response, accumulating content and tool calls
     * into the same `{role, content, tool_calls}` shape produced by the
     * non-streaming path. Mirrors Python's `_handle_stream` and TS's
     * `_handle_stream` so downstream code is identical regardless of mode.
     *
     * @param StreamResponse<CreateStreamedResponse> $stream
     * @return array{role: string, content: ?string, tool_calls: ?array}
     */
    protected function handleStream(StreamResponse $stream): array
    {
        $contentParts = [];
        $toolCallsByIndex = [];

        /** @var CreateStreamedResponse $chunk */
        foreach ($stream as $chunk) {
            $choice = $chunk->choices[0] ?? null;
            if ($choice === null) {
                continue;
            }
            $delta = $choice->delta;

            if ($delta->content !== null && $delta->content !== '') {
                $contentParts[] = $delta->content;
            }

            foreach ($delta->toolCalls as $tc) {
                $idx = $tc->index ?? count($toolCallsByIndex);
                if (!isset($toolCallsByIndex[$idx])) {
                    $toolCallsByIndex[$idx] = [
                        'id' => $tc->id ?? '',
                        'type' => 'function',
                        'function' => ['name' => '', 'arguments' => ''],
                    ];
                }
                $entry = &$toolCallsByIndex[$idx];
                if ($tc->id !== null && $tc->id !== '') {
                    $entry['id'] = $tc->id;
                }
                if ($tc->function->name !== null && $tc->function->name !== '') {
                    $entry['function']['name'] .= $tc->function->name;
                }
                if ($tc->function->arguments !== '') {
                    $entry['function']['arguments'] .= $tc->function->arguments;
                }
                unset($entry);
            }
        }

        $message = ['role' => 'assistant', 'content' => null, 'tool_calls' => null];
        if ($contentParts !== []) {
            $message['content'] = implode('', $contentParts);
        }
        if ($toolCallsByIndex !== []) {
            ksort($toolCallsByIndex);
            $message['tool_calls'] = array_values($toolCallsByIndex);
        }
        return $message;
    }

    /**
     * Convert a non-streaming CreateResponse into the harness's array shape.
     *
     * @return array{role: string, content: ?string, tool_calls: ?array}
     */
    protected function handleResponseObject(CreateResponse $response): array
    {
        $msg = $response->choices[0]->message ?? null;
        $content = $msg?->content;
        $toolCalls = null;
        if ($msg !== null && $msg->toolCalls !== []) {
            $toolCalls = array_map(static function ($tc): array {
                return [
                    'id' => $tc->id,
                    'type' => $tc->type,
                    'function' => [
                        'name' => $tc->function->name,
                        'arguments' => $tc->function->arguments,
                    ],
                ];
            }, $msg->toolCalls);
        }
        return ['role' => 'assistant', 'content' => $content, 'tool_calls' => $toolCalls];
    }

    /**
     * Call the LLM with retry logic.
     *
     * @param array $messages
     * @param array|null $toolsSchema
     * @return array The assistant message
     */
    protected function callLlm(array $messages, ?array $toolsSchema = null): array
    {
        $params = array_merge([
            'model' => $this->model,
            'messages' => $messages,
        ], $this->completionParams);

        if ($toolsSchema !== null && count($toolsSchema) > 0) {
            $params['tools'] = $toolsSchema;
        }

        $lastException = null;
        for ($attempt = 0; $attempt <= $this->maxRetries; $attempt++) {
            try {
                if ($this->stream) {
                    $stream = $this->client->chat()->createStreamed($params);
                    return $this->handleStream($stream);
                }
                $response = $this->client->chat()->create($params);
                return $this->handleResponseObject($response);
            } catch (\Throwable $e) {
                $lastException = $e;
                if ($attempt < $this->maxRetries) {
                    $delay = min(2 ** $attempt, 10);
                    sleep($delay);
                }
            }
        }

        throw new \RuntimeException(
            'LLM call failed after ' . ($this->maxRetries + 1) . ' attempts: ' . $lastException->getMessage(),
            0,
            $lastException,
        );
    }

    /**
     * Run the agent loop.
     *
     * @param array $messages
     * @param array $options
     * @return string
     */
    public function run(array $messages, array $options = []): string
    {
        $messages = array_map(fn($m) => $m, $messages); // shallow copy

        $systemPrompt = $this->buildSystemPrompt($this->system, $options);
        if ($systemPrompt !== null) {
            if (!empty($messages) && ($messages[0]['role'] ?? '') === 'system') {
                $messages[0]['content'] = $systemPrompt;
            } else {
                array_unshift($messages, ['role' => 'system', 'content' => $systemPrompt]);
            }
        }

        $context = new RunContext(agent: $this, turn: 0, metadata: []);
        $this->onRunStart($context);

        for ($turn = 0; $turn < $this->maxTurns; $turn++) {
            $context->turn = $turn;
            $assistantMsg = $this->callLlm($messages);
            $result = $this->handleResponse($assistantMsg, $context);

            if ($result === null) {
                $messages[] = $assistantMsg;
                $content = $assistantMsg['content'] ?? '';
                $this->onRunEnd($context);
                return $content;
            }

            $messages[] = $result;
            if (empty($result['tool_calls'])) {
                $this->onRunEnd($context);
                return $result['content'] ?? '';
            }
        }

        $this->onRunEnd($context);
        return end($messages)['content'] ?? '';
    }
}

File: src/python/base_agent.py
from __future__ import annotations

import asyncio
import json
import logging
from copy import deepcopy
from dataclasses import dataclass, field
from typing import Any

import litellm

logger = logging.getLogger(__name__)


@dataclass
class RunContext:
    agent: Any
    turn: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)


class BaseAgent:
    def __init__(
        self,
        model: str,
        system: str | None = None,
        max_turns: int = 20,
        max_retries: int = 2,
        stream: bool = True,
        **litellm_kwargs: Any,
    ) -> None:
        self.model = model
        self.system = system
        self.max_turns = max_turns
        self.max_retries = max_retries
        self.stream = stream
        self.litellm_kwargs = litellm_kwargs

    async def _build_system_prompt(self, base_prompt: str | None, context: Any) -> str | None:
        return base_prompt

    async def _on_run_start(self, context: RunContext) -> None:
        pass

    async def _on_run_end(self, context: RunContext) -> None:
        pass

    async def _handle_stream(self, stream: Any) -> dict[str, Any]:
        content_parts: list[str] = []
        tool_calls_map: dict[int, dict] = {}

        async for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta is None:
                continue
            if delta.content:
                content_parts.append(delta.content)
            if delta.tool_calls:
                for tc in delta.tool_calls:
                    idx = tc.index
                    if idx not in tool_calls_map:
                        tool_calls_map[idx] = {
                            "id": tc.id or "",
                            "type": "function",
                            "function": {"name": "", "arguments": ""},
                        }
                    entry = tool_calls_map[idx]
                    if tc.id:
                        entry["id"] = tc.id
                    if tc.function:
                        if tc.function.name:
                            entry["function"]["name"] += tc.function.name
                        if tc.function.arguments:
                            entry["function"]["arguments"] += tc.function.arguments

        message: dict[str, Any] = {"role": "assistant"}
        if content_parts:
            message["content"] = "".join(content_parts)
        if tool_calls_map:
            message["tool_calls"] = [tool_calls_map[i] for i in sorted(tool_calls_map)]
        return message

    async def _handle_response(self, response: dict[str, Any], context: RunContext) -> dict[str, Any] | None:
        return response

    async def _call_llm(self, messages: list[dict], tools_schema: list[dict] | None = None) -> dict[str, Any]:
        for attempt in range(self.max_retries + 1):
            try:
                kwargs: dict[str, Any] = {
                    "model": self.model,
                    "messages": messages,
                    **self.litellm_kwargs,
                }
                if tools_schema:
                    kwargs["tools"] = tools_schema

                if self.stream:
                    kwargs["stream"] = True
                    resp = await litellm.acompletion(**kwargs)
                    return await self._handle_stream(resp)
                else:
                    resp = await litellm.acompletion(**kwargs)
                    msg = resp.choices[0].message
                    return {"role": "assistant", "content": msg.content, "tool_calls": getattr(msg, "tool_calls", None)}
            except Exception as e:
                if attempt < self.max_retries:
                    delay = min(2 ** attempt, 10)
                    logger.warning("LLM call failed (attempt %d): %s", attempt + 1, e)
                    await asyncio.sleep(delay)
                else:
                    raise RuntimeError(f"LLM call failed after {self.max_retries + 1} attempts: {e}") from e
        raise RuntimeError("Unreachable")

    async def run(self, messages: list[dict], **kwargs: Any) -> str:
        messages = deepcopy(messages)
        system_prompt = await self._build_system_prompt(self.system, kwargs)

        if system_prompt:
            if messages and messages[0].get("role") == "system":
                messages[0]["content"] = system_prompt
            else:
                messages.insert(0, {"role": "system", "content": system_prompt})

        context = RunContext(agent=self, turn=0, metadata={})
        await self._on_run_start(context)

        for turn in range(self.max_turns):
            context.turn = turn
            assistant_msg = await self._call_llm(messages)
            result = await self._handle_response(assistant_msg, context)

            if result is None:
                messages.append(assistant_msg)
                content = assistant_msg.get("content", "")
                await self._on_run_end(context)
                return content or ""

            messages.append(result)
            if not result.get("tool_calls"):
                await self._on_run_end(context)
                return result.get("content", "")

        await self._on_run_end(context)
        return messages[-1].get("content", "")
